### PR TITLE
Further improval of controls

### PR DESCRIPTION
**Pull request by otaku42**
Hi @tiagocoutinho . I'm still fairly new to Git and collaboration here on Github, so I thought I'd ask you for your input.

I've worked a bit to further improve controls (see this Draft PR), which happened in the improve_controls branch in [otaku42/v4l2py](https://github.com/otaku42/v4l2py/), and I would like to contribute that work upstream.

While I did my stuff, you have committed a nice bunch of changes. This results in  improve_controls being [35 commits ahead](https://github.com/otaku42/v4l2py/compare/tiagocoutinho:v4l2py:master...improve_controls), [65 commits behind](https://github.com/otaku42/v4l2py/compare/improve_controls...tiagocoutinho:v4l2py:master) of your master branch. Merging is not easily possible due to merge conflicts that need to be resolved. So far, so good - but what is the best way to do this?

The "sync fork" tool on Github would discard my 35 commits "to make [the improve_controls] branch match the upstream repository". Although it just might be the explanation what happens during a merge anyway, I don't feel comfortable with trying what happens next, as I don't want to mess the repository/branch.

I'd rather resort to merging things locally and then pushing the results to Github, and I see the following options for that:

1. otaku42/4l2py:master is tracking tiagocoutinho/v4l2py:master. I could merge the result of your 65 commits from otaku42/4l2py:master into otaku42/v4l2py:improve_controls, then send the PR. If you accept the PR, the changes from my improve_controls branch will go into tiagocoutinho/v4l2py:master, from where they will then be synced into otaku42/v4l2py:master.
2. Similar to 1., but merge to otaku42/v4l2py:improve_controls from tiagocoutinho/v4l2py:master rather than from otaku42/v4l2py:master; code-wise the result would be the same, but I'm not sure if it makes a difference on the Git/Github level - or if it is actually even possible.
3. Merge otaku42/v4l2py:improve_controls into otak42/v4l2py:master, and create the PR from there. In contrast to 1, otaku42/v4l2py:master would be ahead of tiagocoutinho/v4l2py:master until you accept and merge the PR.

Assuming that you're willing to accept the changes, which of these options would you consider to be the best way to contribute them? Or are there other, better options that I've missed? I'd appreciate any suggestion.

### CODE REVIEW

**Comment by otaku42** [comment]
It turns out that my option 1 matches what the Pro Git book [describes](https://git-scm.com/book/en/v2/GitHub-Contributing-to-a-Project#ch06-github_flow) being "the GitHub workflow". I think I'll try that path and see where it ends :-)

**Comment by otaku42** [comment]
It turns out that my option 1 matches what the Pro Git book [describes](https://git-scm.com/book/en/v2/GitHub-Contributing-to-a-Project#ch06-github_flow) being "the GitHub workflow". I think I'll try that path and see where and how it ends :-)

File: v4l2py/device.py [hunk 1]
```diff
@@ -1076,13 +1102,13 @@ def set_to_maximum(self):
 
 
 class IntegerControl(BaseNumericControl):
-    lower_bound = -2 ** 31
-    upper_bound = 2 ** 31
+    lower_bound = -(2**31)
+    upper_bound = 2**31
```
Comment (tiagocoutinho) [inline v4l2py/device.py:1106]: isn't max int 32bits: `2**31 - 1` ?

File: v4l2py/device.py [hunk 2]
```diff
@@ -1076,13 +1102,13 @@ def set_to_maximum(self):
...
 
 
 class Integer64Control(BaseNumericControl):
-    lower_bound = -2 ** 63
-    upper_bound = 2 ** 63
+    lower_bound = -(2**63)
+    upper_bound = 2**63
```
Comment (tiagocoutinho) [inline v4l2py/device.py:1111]: isn't max int 64bits: `2**63 - 1` ?

File: v4l2py/device.py (hunk 1 above)
Comment (otaku42) [reply v4l2py/device.py]: Indeed, thanks, good catch. Fixed in [d485978](https://github.com/tiagocoutinho/v4l2py/pull/42/commits/d485978202c0a20925690d1b9707a1351bf1e6e7).

**Comment by otaku42** [comment]
Ok, with a9d5f3a I think I now have covered everything I originally wanted to have in this PR.

To summarize, the idea is to provide a more pythonic API for working with a device's controls. See otaku42/v4l2py#12 for the initial motivation, and otaku42/v4l2py#15 for some kind of rationale. A basic comparison of the old vs. the new API is given in [README.md](https://github.com/otaku42/v4l2py/tree/improve_controls#improved-device-controls), and with [v4l2py-ctl.py](https://github.com/otaku42/v4l2py/blob/improve_controls/examples/v4l2py-ctl.py) there is an example of how to use the new API.

Please let me know if you have any concerns, suggestions or questions.

**Comment by tiagocoutinho** [comment]
Thanks for these awesome improvements @otaku42! 🤩 

I tried the examples/web with your branch but they fail with:
```
jinja2.exceptions.UndefinedError: 'v4l2py.device.IntegerControl object' has no attribute 'info'
```

would it be possible to include a fix? Thanks in advance

**Comment by otaku42** [comment]
> ``` jinja2.exceptions.UndefinedError: 'v4l2py.device.IntegerControl object' has no attribute 'info'```

Oops. Thanks for pointing that out, I'll fix it.

I didn't test the examples by now, to be honest, but was about to do that soon. I didn't expect they make use of the controls API. 

The info-attribute of controls has been renamed to _info, to signal that while it can be accessed it's on ones own risk. Sub-attributes of info are made available as top-level attributes if and where it makes sense. For example, minimum or maximum may or may not be available depending on the control type - see otaku42/v4l2py#12, which is what triggered this whole controls improvement thing :-)

**Comment by otaku42** [comment]
I've taken the lazy approach to fix the issue you have reported; see [a98df12](https://github.com/tiagocoutinho/v4l2py/pull/42/commits/a98df12f1cdc84ec8d4ba169c9e1d86bcba2c880) . 

While working on it, I came across some issues:

- otaku42/v4l2py#24
- otaku42/v4l2py#25
- otaku42/v4l2py#26

and had some ideas for slight enhancements for  `examples/web/async.py` :

- otaku42/v4l2py#27

I didn't want to cram these things into the PR, but I'd try to fix them at a later time, after this PR has been merged.

**Comment by tiagocoutinho** [comment]
Thanks for the fix and the extra work to keep example working with python 3.7.
I really appreciate you being so thorough. I agree the issues you point out and the proposal for improvements should be addressed in other issues/PRs

It is an honor to accept your PR

**Review by tiagocoutinho — APPROVED** [review]
LGTM! Merging

**Comment by otaku42** [comment]
> It is an honor to accept your PR

It's my pleasure, and I'm glad that I can contribute a tiny bit to this great project. Thanks for the great work - and for accepting my contributions :-)

**Pull request merged**